README.md
## RPCircularProgress

[![Coverage Status](https://coveralls.io/repos/github/iwasrobbed/RPCircularProgress/badge.svg?branch=master)](https://coveralls.io/github/iwasrobbed/RPCircularProgress?branch=master)
[![Build Status](https://travis-ci.org/iwasrobbed/RPCircularProgress.svg?branch=master)](https://travis-ci.org/iwasrobbed/RPCircularProgress)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/iwasrobbed/RPCircularProgress/blob/master/LICENSE)
[![CocoaPods](https://img.shields.io/cocoapods/v/RPCircularProgress.svg?maxAge=2592000)]()
[![Swift](https://img.shields.io/badge/language-Swift-blue.svg)](https://swift.org)

⚠️ To use with Swift 2.3 please ensure you are using == 0.2.3 ⚠️

⚠️ To use with Swift 3.x please ensure you are using >= 0.3.0 ⚠️

⚠️ To use with Swift 4.x please ensure you are using >= 0.4.0 ⚠️

⚠️ To use with Swift 4.2 please ensure you are using >= 0.4.1 ⚠️

UIView subclass written in Swift to show circular progress.

![Example Usage](ohhai.gif)

Is your app using it? [Let me know!](mailto:rob@desideratalabs.co)

Please see the included example app for sample usage.

### Styles

* Indeterminate (spins infinitely)
* Normal (set progress between 0.0 and 1.0)
* Mixed (can make it indeterminate, but also animate progress to 1.0)

### Public API

```swift
/**
 Enables or disables the indeterminate (spinning) animation

 - parameter enabled:    Whether or not to enable the animation (defaults to `true`)
 - parameter completion: An optional closure to execute after the animation completes
 */
open func enableIndeterminate(_ enabled: Bool = true, completion: CompletionBlock? = nil) {}
```

```swift
/**
 Updates the progress bar to the given value with the optional properties

 - parameter progress:     The progress to update to, pinned between `0` and `1`
 - parameter animated:     Whether or not the update should be animated (defaults to `true`)
 - parameter initialDelay: Sets an initial delay before the animation begins
 - parameter duration:     Sets the overal duration that the animation should complete within
 - parameter completion:   An optional closure to execute after the animation completes
 */
open func updateProgress(_ progress: CGFloat, animated: Bool = true, initialDelay: CFTimeInterval = 0, duration: CFTimeInterval? = nil, completion: CompletionBlock? = nil) {}
```

### Properties

Note: Most properties below are `@IBInspectable`, but I don't use Interface Builder personally so let me know if you see any issues.

`var trackTintColor: UIColor`
* The color of the empty progress track (gets drawn over)

`var progressTintColor: UIColor`
* The color of the progress bar

`var innerTintColor: UIColor?`
* The color the notched out circle within the progress area (if there is one)

`var roundedCorners: Bool`
* Sets whether or not the corners of the progress bar should be rounded

`var thicknessRatio: CGFloat`
* Sets how thick the progress bar should be (pinned between `0.01` and `1`)

`var clockwiseProgress: Bool`
* Sets whether or not the animation should be clockwise

`var timingFunction: CAMediaTimingFunction`
* A timing function defining the pacing of the animation. Defaults to ease in, ease out.

`var progress: CGFloat`
* Getter for the current progress (not observed from any active animations)

`var indeterminateProgress: CGFloat`
* Sets how much of the progress bar should be filled during an indeterminate animation, pinned between `0.05` and `0.9`
* **Note:** This can be overriden / animated from by using updateProgress(...)

`var indeterminateDuration: CFTimeInterval`
* Controls the speed at which the indeterminate progress bar animates

### Supports
Swift, ARC & iOS 8+, Autolayout or springs and struts

### A little help from my friends
Please feel free to fork and create a pull request for bug fixes or improvements, being sure to maintain the general coding style, adding tests, and adding comments as necessary.

### Credit
This library is effectively a Swift port of [DACircularProgress](https://github.com/danielamitay/DACircularProgress) with some minor changes to the API, so it should be fairly easy to act as a replacement. I really loved that library but unfortunately it wasn't being maintained.


CHANGELOG.md
# Change Log

## [v0.5.0](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.5.0) (2019-04-06)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.4.1...v0.5.0)

**Closed issues:**

- s.swift\_version = '4.2' missing from podspec [\#34](https://github.com/iwasrobbed/RPCircularProgress/issues/34)
- Xcode 10.2 / Swift 5 support [\#36](https://github.com/iwasrobbed/RPCircularProgress/issues/36)

**Merged pull requests:**

- Swift/5.0 [\#37](https://github.com/iwasrobbed/RPCircularProgress/pull/37) ([alex-taffe](https://github.com/alex-taffe))
- Added swift version to podspec [\#35](https://github.com/iwasrobbed/RPCircularProgress/pull/35) ([kamrankhan07](https://github.com/kamrankhan07))

## [v0.4.1](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.4.1) (2018-10-10)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.4.0...v0.4.1)

**Closed issues:**

- libswiftXCTest.dylib   Reason: image not found [\#32](https://github.com/iwasrobbed/RPCircularProgress/issues/32)

**Merged pull requests:**

- Swift 4.2 Migration [\#33](https://github.com/iwasrobbed/RPCircularProgress/pull/33) ([alex-taffe](https://github.com/alex-taffe))
- Remove use of casting to implicitly unwrapped optional [\#30](https://github.com/iwasrobbed/RPCircularProgress/pull/30) ([axtonpitt](https://github.com/axtonpitt))

## [v0.4.0](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.4.0) (2017-10-07)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.3.1...v0.4.0)

**Implemented enhancements:**

- Swift 4.0 support [\#27](https://github.com/iwasrobbed/RPCircularProgress/pull/27) ([iwasrobbed](https://github.com/iwasrobbed))

**Closed issues:**

- Swift 4 support 🙏 [\#26](https://github.com/iwasrobbed/RPCircularProgress/issues/26)

## [v0.3.1](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.3.1) (2017-05-16)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.3.0...v0.3.1)

**Closed issues:**

- UpdateProgress not available from Objective C [\#23](https://github.com/iwasrobbed/RPCircularProgress/issues/23)
- Dot Appears every time updateProgress is called. [\#21](https://github.com/iwasrobbed/RPCircularProgress/issues/21)
- just a noob question [\#19](https://github.com/iwasrobbed/RPCircularProgress/issues/19)
- Swift 3 to Objc Generated interface  [\#18](https://github.com/iwasrobbed/RPCircularProgress/issues/18)
- CocoaPods support. [\#17](https://github.com/iwasrobbed/RPCircularProgress/issues/17)
- Question: Start Progress [\#16](https://github.com/iwasrobbed/RPCircularProgress/issues/16)
- Swift 3 and iOS 10 Support [\#10](https://github.com/iwasrobbed/RPCircularProgress/issues/10)

**Merged pull requests:**

- Fix Swift 3.1 compile warnings [\#24](https://github.com/iwasrobbed/RPCircularProgress/pull/24) ([Magnat12](https://github.com/Magnat12))

## [v0.3.0](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.3.0) (2016-09-19)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.2.3...v0.3.0)

**Merged pull requests:**

- Swift 3.0 support [\#15](https://github.com/iwasrobbed/RPCircularProgress/pull/15) ([iwasrobbed](https://github.com/iwasrobbed))
- Swift 3.0 support [\#13](https://github.com/iwasrobbed/RPCircularProgress/pull/13) ([iwasrobbed](https://github.com/iwasrobbed))
- Swift 2.3 support [\#12](https://github.com/iwasrobbed/RPCircularProgress/pull/12) ([iwasrobbed](https://github.com/iwasrobbed))
- Swift 2.3 support [\#11](https://github.com/iwasrobbed/RPCircularProgress/pull/11) ([iwasrobbed](https://github.com/iwasrobbed))

## [v0.2.3](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.2.3) (2016-09-19)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.1.2...v0.2.3)

**Implemented enhancements:**

- Objective-C support [\#9](https://github.com/iwasrobbed/RPCircularProgress/issues/9)

**Merged pull requests:**

- Update README [\#7](https://github.com/iwasrobbed/RPCircularProgress/pull/7) ([iwasrobbed](https://github.com/iwasrobbed))
- Code coverage  [\#6](https://github.com/iwasrobbed/RPCircularProgress/pull/6) ([iwasrobbed](https://github.com/iwasrobbed))
- Code coverage [\#5](https://github.com/iwasrobbed/RPCircularProgress/pull/5) ([iwasrobbed](https://github.com/iwasrobbed))
- Readme backfill  [\#4](https://github.com/iwasrobbed/RPCircularProgress/pull/4) ([iwasrobbed](https://github.com/iwasrobbed))
- Access for pod use & allow custom timing functions [\#3](https://github.com/iwasrobbed/RPCircularProgress/pull/3) ([iwasrobbed](https://github.com/iwasrobbed))

## [v0.1.2](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.1.2) (2016-04-08)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.1.1...v0.1.2)

## [v0.1.1](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.1.1) (2016-04-07)
[Full Changelog](https://github.com/iwasrobbed/RPCircularProgress/compare/v0.1...v0.1.1)

**Merged pull requests:**

- Update podspec for lint [\#2](https://github.com/iwasrobbed/RPCircularProgress/pull/2) ([iwasrobbed](https://github.com/iwasrobbed))

## [v0.1](https://github.com/iwasrobbed/RPCircularProgress/tree/v0.1) (2016-04-07)
**Merged pull requests:**

- Adds travis [\#1](https://github.com/iwasrobbed/RPCircularProgress/pull/1) ([iwasrobbed](https://github.com/iwasrobbed))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


Tests/RPCircularProgressTests.swift
//
//  RPCircularProgressTests.swift
//  RPCircularProgressTests
//
//  Created by Rob Phillips on 4/6/16.
//  Copyright © 2016 Glazed Donut, LLC. All rights reserved.
//
//  See LICENSE for full license agreement.
//

import Quick
import Nimble

@testable import RPCircularProgressExample

class RPCircularProgressTests: QuickSpec {
    
    override func spec() {

        let progress = RPCircularProgress()

        describe("properties") { 

            context("colors") {

                // Non-default color
                let color = UIColor.purple

                it("should update track color") {
                    progress.trackTintColor = color
                    expect(progress.trackTintColor) == color
                }

                it("should update progress color") {
                    progress.progressTintColor = color
                    expect(progress.progressTintColor) == color
                }

                it("should have an optional inner area color") {
                    expect(progress.innerTintColor).to(beNil())
                }

                it("should update inner area color") {
                    progress.innerTintColor = color
                    expect(progress.innerTintColor) == color
                }

            }

            context("corners") {

                it("should update rounded") {
                    progress.roundedCorners = false
                    expect(progress.roundedCorners).to(beFalse())

                    progress.roundedCorners = true
                    expect(progress.roundedCorners).to(beTrue())
                }

            }

            context("progress bar") {

                it("should update thickness ratio") {
                    progress.thicknessRatio = 0.42
                    expect(progress.thicknessRatio) == 0.42
                }

                it("should pin thickness ratio") {
                    progress.thicknessRatio = -1
                    expect(progress.thicknessRatio) == 0.01

                    progress.thicknessRatio = 2
                    expect(progress.thicknessRatio) == 1
                }

                it("should update clockwise progress") {
                    progress.clockwiseProgress = false
                    expect(progress.clockwiseProgress).to(beFalse())

                    progress.clockwiseProgress = true
                    expect(progress.clockwiseProgress).to(beTrue())
                }

                it("should have the correct timing function") {
                    let timingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.linear)
                    progress.timingFunction = timingFunction
                    progress.updateProgress(1, duration: 2)

                    let animation = progress.layer.animation(forKey: "progress")
                    expect(animation).toEventuallyNot(beNil())

                    let currentTimingFunction = animation!.value(forKey: "timingFunction") as! CAMediaTimingFunction
                    expect(currentTimingFunction).to(equal(timingFunction))
                }

                it("should return current progress") {
                    progress.updateProgress(0.42, animated: false)
                    expect(progress.progress) == 0.42
                }
                
            }

            context("indeterminate progress bar") {

                it("should update indeterminate progress") {
                    progress.indeterminateProgress = 0.42
                    expect(progress.indeterminateProgress) == 0.42
                }

                it("should pin indeterminate progress") {
                    progress.indeterminateProgress = -1
                    expect(progress.indeterminateProgress) == 0.05

                    progress.indeterminateProgress = 2
                    expect(progress.indeterminateProgress) == 0.9
                }

                it("should update indeterminate duration") {
                    progress.indeterminateDuration = 42
                    expect(progress.indeterminateDuration) == 42
                }

            }

        }

        describe("indeterminate progress") {

            context("enabling or disabling") {

                let key = "indeterminateAnimation"

                it("should enable the animation") {
                    progress.enableIndeterminate()

                    let animation = progress.layer.animation(forKey: key)
                    expect(animation).toEventuallyNot(beNil())
                }

                it("should disable the animation") {
                    progress.enableIndeterminate()
                    var animation = progress.layer.animation(forKey: key)
                    expect(animation).toEventuallyNot(beNil())

                    progress.enableIndeterminate(false)
                    animation = progress.layer.animation(forKey: key)
                    expect(animation).to(beNil())
                }

            }

            context("completion") {

                it("should call the enabling closure upon completion") {
                    waitUntil { done in
                        progress.enableIndeterminate(completion: {
                            expect(true).to(beTrue())
                            done()
                        })
                        progress.enableIndeterminate(false)
                    }
                }

                it("should call the disabling closure upon completion") {
                    progress.enableIndeterminate()
                    waitUntil { done in
                        progress.enableIndeterminate(false, completion: {
                            expect(true).to(beTrue())
                            done()
                        })
                    }
                }
                
            }

        }

        describe("updating progress") {

            let key = "progress"

            context("progress") {

                it("should pin progress") {
                    progress.updateProgress(-1, animated: false)
                    expect(progress.progress) == 0

                    progress.updateProgress(2, animated: false)
                    expect(progress.progress) == 1
                }

                it("should update to the proper amount") {
                    progress.updateProgress(0.42, animated: false)
                    expect(progress.progress).to(beCloseTo(0.42))
                }

            }

            context("animated") {

                it("should add animation if passed as true") {
                    progress.updateProgress(0.42, animated: true)

                    let animation = progress.layer.animation(forKey: key)

                    expect(animation).toEventuallyNot(beNil())
                    expect(progress.progress).toEventually(equal(0.42))
                }

                it("should not add animation if passed as false") {
                    progress.updateProgress(0.42, animated: false)

                    let animation = progress.layer.animation(forKey: key)

                    expect(animation).to(beNil())
                    expect(progress.progress) == 0.42
                }

            }

            context("initialDelay") {

                it("should set the proper value") {
                    let beginTime = NSNumber(floatLiteral: CACurrentMediaTime() + 42)
                    progress.updateProgress(0.42, initialDelay: 42)

                    let animation = progress.layer.animation(forKey: key)
                    expect(animation).toEventuallyNot(beNil())

                    let initialDelay = animation?.value(forKey: "beginTime") as? NSNumber
                    expect(initialDelay?.intValue) == beginTime.intValue
                }

            }

            context("duration") {

                it("should set the proper value") {
                    progress.updateProgress(0.42, duration: 42)

                    let animation = progress.layer.animation(forKey: key)
                    expect(animation).toEventuallyNot(beNil())

                    let duration = animation?.value(forKey: "duration") as? Int
                    expect(duration) == 42
                }

            }

            context("completion") {

                it("should call the closure upon completion") {
                    waitUntil { done in
                        progress.updateProgress(0.42, duration: 42, completion: {
                            expect(true).to(beTrue())
                            done()
                        })
                    }
                }

            }

        }

    }

}


Source/RPCircularProgress.swift
//
//  RPCircularProgress.swift
//  RPCircularProgress
//
//  Created by Rob Phillips on 4/5/16.
//  Copyright © 2016 Glazed Donut, LLC. All rights reserved.
//
//  See LICENSE for full license agreement.
//

import UIKit

open class RPCircularProgress: UIView {

    // MARK: - Completion

    public typealias CompletionBlock = () -> Void

    // MARK: - Public API

    /**
      The color of the empty progress track (gets drawn over)
    */
    @IBInspectable open var trackTintColor: UIColor {
        get {
            return progressLayer.trackTintColor
        }
        set {
            progressLayer.trackTintColor = newValue
            progressLayer.setNeedsDisplay()
        }
    }

    /**
      The color of the progress bar
     */
    @IBInspectable open var progressTintColor: UIColor {
        get {
            return progressLayer.progressTintColor
        }
        set {
            progressLayer.progressTintColor = newValue
            progressLayer.setNeedsDisplay()
        }
    }

    /**
      The color the notched out circle within the progress area (if there is one)
     */
    @IBInspectable open var innerTintColor: UIColor? {
        get {
            return progressLayer.innerTintColor
        }
        set {
            progressLayer.innerTintColor = newValue
            progressLayer.setNeedsDisplay()
        }
    }

    /**
      Sets whether or not the corners of the progress bar should be rounded
     */
    @IBInspectable open var roundedCorners: Bool {
        get {
            return progressLayer.roundedCorners
        }
        set {
            progressLayer.roundedCorners = newValue
            progressLayer.setNeedsDisplay()
        }
    }

    /**
      Sets how thick the progress bar should be (pinned between `0.01` and `1`)
     */
    @IBInspectable open var thicknessRatio: CGFloat {
        get {
            return progressLayer.thicknessRatio
        }
        set {
            progressLayer.thicknessRatio = pin(newValue, minValue: 0.01, maxValue: 1)
            progressLayer.setNeedsDisplay()
        }
    }

    /**
      Sets whether or not the animation should be clockwise
     */
    @IBInspectable open var clockwiseProgress: Bool {
        get {
            return progressLayer.clockwiseProgress
        }
        set {
            progressLayer.clockwiseProgress = newValue
            progressLayer.setNeedsDisplay()
        }
    }

    /**
      A timing function defining the pacing of the animation. Defaults to ease in, ease out.
     */
    open var timingFunction: CAMediaTimingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.easeInEaseOut)

    /**
      Getter for the current progress (not observed from any active animations)
     */
    @IBInspectable open var progress: CGFloat {
        get {
            return progressLayer.progress
        }
    }

    /**
      Sets how much of the progress bar should be filled during an indeterminate animation, pinned between `0.05` and `0.9`
     
      **Note:** This can be overriden / animated from by using updateProgress(...)
     */
    @IBInspectable open var indeterminateProgress: CGFloat {
        get {
            return progressLayer.indeterminateProgress
        }
        set {
            progressLayer.indeterminateProgress = pin(newValue, minValue: 0.05, maxValue: 0.9)
        }
    }

    /**
      Controls the speed at which the indeterminate progress bar animates
     */
    @IBInspectable open var indeterminateDuration: CFTimeInterval = Defaults.indeterminateDuration

    // MARK: - Custom Base Layer

    fileprivate var progressLayer: ProgressLayer! {
        get {
            return (layer as! ProgressLayer)
        }
    }

    open override class var layerClass : AnyClass {
        return ProgressLayer.self
    }

    // Lifecycle

    /**
     Default initializer for the class

     - returns: A configured instance of self
     */
    required public init() {
        super.init(frame: CGRect(x: 0, y: 0, width: 40, height: 40))

        setupDefaults()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)

        setupDefaults()
    }

    open override func didMoveToWindow() {
        super.didMoveToWindow()

        if let window = window {
            progressLayer.contentsScale = window.screen.scale
            progressLayer.setNeedsDisplay()
        }
    }

    // MARK: - Indeterminate

    /**
     Enables or disables the indeterminate (spinning) animation

     - parameter enabled:    Whether or not to enable the animation (defaults to `true`)
     - parameter completion: An optional closure to execute after the animation completes
     */
    open func enableIndeterminate(_ enabled: Bool = true, completion: CompletionBlock? = nil) {
        if let animation = progressLayer.animation(forKey: AnimationKeys.indeterminate) {
            // Check if there are any closures to execute on the existing animation
            if let block = animation.value(forKey: AnimationKeys.completionBlock) as? CompletionBlockObject {
                block.action()
            }
            progressLayer.removeAnimation(forKey: AnimationKeys.indeterminate)

            // And notify of disabling completion
            completion?()
        }

        guard enabled else { return }

        addIndeterminateAnimation(completion)
    }

    // MARK: - Progress

    /**
     Updates the progress bar to the given value with the optional properties

     - parameter progress:     The progress to update to, pinned between `0` and `1`
     - parameter animated:     Whether or not the update should be animated (defaults to `true`)
     - parameter initialDelay: Sets an initial delay before the animation begins
     - parameter duration:     Sets the overal duration that the animation should complete within
     - parameter completion:   An optional closure to execute after the animation completes
     */
    open func updateProgress(_ progress: CGFloat, animated: Bool = true, initialDelay: CFTimeInterval = 0, duration: CFTimeInterval? = nil, completion: CompletionBlock? = nil) {
        let pinnedProgress = pin(progress)
        if animated {

            // Get duration
            let animationDuration: CFTimeInterval
            if let duration = duration, duration != 0 {
                animationDuration = duration
            } else {
                // Same duration as UIProgressView animation
                animationDuration = CFTimeInterval(fabsf(Float(self.progress) - Float(pinnedProgress)))
            }

            // Get current progress (to avoid jumpy behavior)
            // Basic animations have their value reset to the original once the animation is finished
            // since only the presentation layer is animating
            var currentProgress: CGFloat = 0
            if let presentationLayer = progressLayer.presentation() as? ProgressLayer {
                currentProgress = presentationLayer.progress
            }
            progressLayer.progress = currentProgress

            progressLayer.removeAnimation(forKey: AnimationKeys.progress)
            animate(progress, currentProgress: currentProgress, initialDelay: initialDelay, duration: animationDuration, completion: completion)
        } else {
            progressLayer.removeAnimation(forKey: AnimationKeys.progress)

            progressLayer.progress = pinnedProgress
            progressLayer.setNeedsDisplay()

            completion?()
        }
    }
}

// MARK: - Private API

private extension RPCircularProgress {

    // MARK: - Defaults

    func setupDefaults() {
        progressLayer.trackTintColor = Defaults.trackTintColor
        progressLayer.progressTintColor = Defaults.progressTintColor
        progressLayer.innerTintColor = nil
        backgroundColor = Defaults.backgroundColor
        progressLayer.thicknessRatio = Defaults.thicknessRatio
        progressLayer.roundedCorners = Defaults.roundedCorners
        progressLayer.clockwiseProgress = Defaults.clockwiseProgress
        indeterminateDuration = Defaults.indeterminateDuration
        progressLayer.indeterminateProgress = Defaults.indeterminateProgress
    }

    // MARK: - Progress

    // Pin certain values between 0.0 and 1.0
    func pin(_ value: CGFloat, minValue: CGFloat = 0, maxValue: CGFloat = 1) -> CGFloat {
        return min(max(value, minValue), maxValue)
    }

    func animate(_ pinnedProgress: CGFloat, currentProgress: CGFloat, initialDelay: CFTimeInterval, duration: CFTimeInterval, completion: CompletionBlock?) {
        let animation = CABasicAnimation(keyPath: AnimationKeys.progress)
        animation.duration = duration
        animation.timingFunction = timingFunction
        animation.fromValue = currentProgress
        animation.fillMode = CAMediaTimingFillMode.forwards
        animation.isRemovedOnCompletion = false
        animation.toValue = pinnedProgress
        animation.beginTime = CACurrentMediaTime() + initialDelay
        animation.delegate = self
        if let completion = completion {
            let completionObject = CompletionBlockObject(action: completion)
            animation.setValue(completionObject, forKey: AnimationKeys.completionBlock)
        }
        progressLayer.add(animation, forKey: AnimationKeys.progress)
    }

    // MARK: - Indeterminate 

    func addIndeterminateAnimation(_ completion: CompletionBlock?) {
        guard progressLayer.animation(forKey: AnimationKeys.indeterminate) == nil else { return }

        let animation = CABasicAnimation(keyPath: AnimationKeys.transformRotation)
        animation.byValue = clockwiseProgress ? 2 * Double.pi : -2 * Double.pi
        animation.duration = indeterminateDuration
        animation.repeatCount = Float.infinity
        animation.isRemovedOnCompletion = false
        progressLayer.progress = indeterminateProgress
        if let completion = completion {
            let completionObject = CompletionBlockObject(action: completion)
            animation.setValue(completionObject, forKey: AnimationKeys.completionBlock)
        }
        progressLayer.add(animation, forKey: AnimationKeys.indeterminate)
    }

    // Completion

    class CompletionBlockObject: NSObject {
        var action: CompletionBlock

        required init(action: @escaping CompletionBlock) {
            self.action = action
        }
    }

    // MARK: - Private Classes / Structs

    class ProgressLayer: CALayer {
        @NSManaged var trackTintColor: UIColor
        @NSManaged var progressTintColor: UIColor
        @NSManaged var innerTintColor: UIColor?

        @NSManaged var roundedCorners: Bool
        @NSManaged var clockwiseProgress: Bool
        @NSManaged var thicknessRatio: CGFloat

        @NSManaged var indeterminateProgress: CGFloat
        // This needs to have a setter/getter for it to work with CoreAnimation
        @NSManaged var progress: CGFloat

        override class func needsDisplay(forKey key: String) -> Bool {
            return key == AnimationKeys.progress ? true : super.needsDisplay(forKey: key)
        }

        override func draw(in ctx: CGContext) {
            let rect = bounds
            let centerPoint = CGPoint(x: rect.size.width / 2, y: rect.size.height / 2)
            let radius = min(rect.size.height, rect.size.width) / 2

            let progress: CGFloat = min(self.progress, CGFloat(1 - Float.ulpOfOne))
            var radians: CGFloat = 0
            if clockwiseProgress {
                radians = CGFloat((Double(progress) * 2 * Double.pi) - (Double.pi / 2))
            } else {
                radians = CGFloat(3 * (Double.pi / 2) - (Double(progress) * 2 * Double.pi))
            }

            func fillTrack() {
                ctx.setFillColor(trackTintColor.cgColor)
                let trackPath = CGMutablePath()
                trackPath.move(to: centerPoint)
                trackPath.addArc(center: centerPoint, radius: radius, startAngle: CGFloat(2 * Double.pi), endAngle: 0, clockwise: true)
                trackPath.closeSubpath()
                ctx.addPath(trackPath)
                ctx.fillPath()
            }

            func fillProgressIfNecessary() {
                if progress == 0.0 {
                    return
                }

                func fillProgress() {
                    ctx.setFillColor(progressTintColor.cgColor)
                    let progressPath = CGMutablePath()
                    progressPath.move(to: centerPoint)
                    progressPath.addArc(center: centerPoint, radius: radius, startAngle: CGFloat(3 * (Double.pi / 2)), endAngle: radians, clockwise: !clockwiseProgress)
                    progressPath.closeSubpath()
                    ctx.addPath(progressPath)
                    ctx.fillPath()
                }

                func roundCornersIfNecessary() {
                    if !roundedCorners {
                        return
                    }

                    let pathWidth = radius * thicknessRatio
                    let xOffset = radius * (1 + ((1 - (thicknessRatio / 2)) * CGFloat(cosf(Float(radians)))))
                    let yOffset = radius * (1 + ((1 - (thicknessRatio / 2)) * CGFloat(sinf(Float(radians)))))
                    let endpoint = CGPoint(x: xOffset, y: yOffset)

                    let startEllipseRect = CGRect(x: centerPoint.x - pathWidth / 2, y: 0, width: pathWidth, height: pathWidth)
                    ctx.addEllipse(in: startEllipseRect)
                    ctx.fillPath()

                    let endEllipseRect = CGRect(x: endpoint.x - pathWidth / 2, y: endpoint.y - pathWidth / 2, width: pathWidth, height: pathWidth)
                    ctx.addEllipse(in: endEllipseRect)
                    ctx.fillPath()
                }

                fillProgress()
                roundCornersIfNecessary()
            }

            func notchCenterCircle() {
                ctx.setBlendMode(.clear)
                let innerRadius = radius * (1 - thicknessRatio)
                let clearRect = CGRect(x: centerPoint.x - innerRadius, y: centerPoint.y - innerRadius, width: innerRadius * 2, height: innerRadius * 2)
                ctx.addEllipse(in: clearRect)
                ctx.fillPath()

                func fillInnerTintIfNecessary() {
                    if let innerTintColor = innerTintColor {
                        ctx.setBlendMode(.normal)
                        ctx.setFillColor(innerTintColor.cgColor)
                        ctx.addEllipse(in: clearRect)
                        ctx.fillPath()
                    }
                }

                fillInnerTintIfNecessary()
            }

            fillTrack()
            fillProgressIfNecessary()
            notchCenterCircle()
        }
    }

    struct Defaults {
        static let trackTintColor = UIColor(white: 1.0, alpha: 0.3)
        static let progressTintColor = UIColor.white
        static let backgroundColor = UIColor.clear

        static let progress: CGFloat = 0
        static let thicknessRatio: CGFloat = 0.3
        static let roundedCorners = true
        static let clockwiseProgress = true
        static let indeterminateDuration: CFTimeInterval = 1.0
        static let indeterminateProgress: CGFloat = 0.3
    }

    struct AnimationKeys {
        static let indeterminate = "indeterminateAnimation"
        static let progress = "progress"
        static let transformRotation = "transform.rotation"
        static let completionBlock = "completionBlock"
        static let toValue = "toValue"
    }

}

// MARK: - Animation Delegate

extension RPCircularProgress: CAAnimationDelegate {

    public func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
        let completedValue = anim.value(forKey: AnimationKeys.toValue)
        if let completedValue = completedValue as? CGFloat {
            progressLayer.progress = completedValue
        }

        if let block = anim.value(forKey: AnimationKeys.completionBlock) as? CompletionBlockObject {
            block.action()
        }
    }

}

